# Unit 12 - Tales from the Crypto

---


## 1. Sentiment Analysis

Use the [newsapi](https://newsapi.org/) to pull the latest news articles for Bitcoin and Ethereum and create a DataFrame of sentiment scores for each coin.

Use descriptive statistics to answer the following questions:
1. Which coin had the highest mean positive score?
2. Which coin had the highest negative score?
3. Which coin had the highest positive score?

In [1]:
# Initial imports
import os
import pandas as pd
import nltk as nltk
from dotenv import load_dotenv
from newsapi import NewsApiClient
from nltk.sentiment.vader import SentimentIntensityAnalyzer

nltk.download('vader_lexicon')
analyzer = SentimentIntensityAnalyzer()
%matplotlib inline

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     /Users/jason/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!

Bad key text.latex.preview in file /Users/jason/opt/anaconda3/envs/pyvizenv/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle, line 123 ('text.latex.preview : False')
You probably need to get an updated matplotlibrc file from
https://github.com/matplotlib/matplotlib/blob/v3.5.1/matplotlibrc.template
or from the matplotlib source distribution

Bad key mathtext.fallback_to_cm in file /Users/jason/opt/anaconda3/envs/pyvizenv/lib/python3.7/site-packages/matplotlib/mpl-data/stylelib/_classic_test.mplstyle, line 155 ('mathtext.fallback_to_cm : True  # When True, use symbols from the Computer Modern')
You probably need to get an updated matplotlibrc file from
https://github.com/matplotlib/matplotlib/blob/v3.5.1/matplotlibrc.template
or from the matplotlib source distribution

Bad key savefig.jpeg_quality in file /U

In [2]:
# Read your api key environment variable
load_dotenv()
api_key = os.getenv("NEWS_API")

In [3]:
# Create a newsapi client
newsapi = NewsApiClient(api_key=api_key)

In [4]:
# Fetch the Bitcoin news articles
btc_articles_dto = newsapi.get_everything(
    q="Bitcoin OR BTC", 
    language="en", 
    sort_by="relevancy")
btc_articles_dto

{'status': 'ok',
 'totalResults': 7897,
 'articles': [{'source': {'id': 'reuters', 'name': 'Reuters'},
   'author': None,
   'title': 'Rouble-crypto trading soars as sanctions hit Russian currency - Reuters',
   'description': 'Trading volumes between the Russian rouble and the Tether cryptocurrency spiked sharply on Monday as the local currency tumbled to a record low on Western sanctions, data shared with Reuters showed.',
   'url': 'https://www.reuters.com/markets/europe/rouble-crypto-trading-soars-sanctions-hit-russian-currency-2022-02-28/',
   'urlToImage': 'https://www.reuters.com/pf/resources/images/reuters/reuters-default.png?d=75',
   'publishedAt': '2022-02-28T17:53:00Z',
   'content': 'LONDON, Feb 28 (Reuters) - Trading volumes between the Russian rouble and the Tether cryptocurrency spiked sharply on Monday as the local currency tumbled to a record low on Western sanctions, data s… [+1745 chars]'},
  {'source': {'id': 'reuters', 'name': 'Reuters'},
   'author': None,
   'ti

In [5]:
# Fetch the Ethereum news articles
eth_articles_dto = newsapi.get_everything(
    q="Ethereum OR ETH", 
    language="en", 
    sort_by="relevancy")
eth_articles_dto

{'status': 'ok',
 'totalResults': 4155,
 'articles': [{'source': {'id': None, 'name': 'Investorplace.com'},
   'author': 'InvestorPlace',
   'title': 'The Market Has Spoken, and It Says Ethereum Is Valuable',
   'description': 'Technical analysis isn’t a perfect tool, but it may point the way for Ethereum Ethereum\xa0(ETH-USD) continues to be a volatile crypto investment. Crypto is volatile by nature — I’m not setting it apart from the asset class. It\xa0 has clear catalysts as well as mu…',
   'url': 'https://investorplace.com/2022/02/the-market-has-spoken-and-it-says-ethereum-matters/',
   'urlToImage': 'https://images.readwrite.com/wp-content/uploads/2021/12/The-Role-of-Crypto-in-The-Fintech-Industry-and-The-Wider-Economy.jpg',
   'publishedAt': '2022-02-17T17:10:12Z',
   'content': 'Technical analysis isnt a perfect tool, but it may point the way for Ethereum\r\nEthereum\xa0(ETH-USD\r\n) continues to be a volatile crypto investment. Crypto is volatile by nature Im not setting it ap

In [6]:
# Create the Bitcoin sentiment scores DataFrame
btc_df = pd.DataFrame.from_dict(btc_articles_dto["articles"])
btc_df.head()

source  \
0      {'id': 'reuters', 'name': 'Reuters'}   
1      {'id': 'reuters', 'name': 'Reuters'}   
2  {'id': 'the-verge', 'name': 'The Verge'}   
3    {'id': None, 'name': 'New York Times'}   
4        {'id': None, 'name': 'Kitco NEWS'}   

                              author  \
0                               None   
1                               None   
2                  Elizabeth Lopatto   
3                       Katie Benner   
4  http://www.facebook.com/kitconews   

                                               title  \
0  Rouble-crypto trading soars as sanctions hit R...   
1    Cryptocurrencies in a time of war - Reuters.com   
2   How Ukrainians are fundraising in cryptocurrency   
3  Justice Dept. Announces Raft of Changes Meant ...   
4  Is crypto winter over? Bitcoin price jumps tow...   

                                         description  \
0  Trading volumes between the Russian rouble and...   
1  Cryptocurrencies have been close to the headli...   
2  Millions of dollars of cryptocurrency have flo...   
3  The moves came a week after the department mad...   
4  <ol><li>Is crypto winter over? Bitcoin price j...   

                                                 url  \
0  https://www.reuters.com/markets/europe/rouble-...   
1  https://www.reuters.com/technology/cryptocurre...   
2  https://www.theverge.com/2022/2/26/22952357/uk...   
3  https://www.nytimes.com/2022/02/17/us/politics...   
4  https://www.kitco.com/news/2022-03-01/Is-crypt...   

                                          urlToImage           publishedAt  \
0  https://www.reuters.com/pf/resources/images/re...  2022-02-28T17:53:00Z   
1  https://www.reuters.com/resizer/o-neidYH62soGC...  2022-03-04T15:01:00Z   
2  https://cdn.vox-cdn.com/thumbor/teEVxppIZ_JTW-...  2022-02-26T20:29:04Z   
3  https://static01.nyt.com/images/2022/02/17/us/...  2022-02-17T23:51:49Z   
4  https://www.kitco.com/news/2022-03-01/images/s...  2022-03-01T22:03:00Z   

                                             content  
0  LONDON, Feb 28 (Reuters) - Trading volumes bet...  
1  LONDON, March 4 (Reuters) - Cryptocurrencies h...  
2  Illustration by James Bareham / The Verge\r\n\...  
3  Even in cyberspace, the Department of Justice ...  
4  (Kitco News) The escalating Russia-Ukraine cri...

In [7]:
# Create the Ethereum sentiment scores DataFrame
eth_df = pd.DataFrame.from_dict(eth_articles_dto["articles"])
eth_df.head()

source  \
0          {'id': None, 'name': 'Investorplace.com'}   
1           {'id': 'the-verge', 'name': 'The Verge'}   
2               {'id': 'reuters', 'name': 'Reuters'}   
3  {'id': 'business-insider', 'name': 'Business I...   
4     {'id': 'the-next-web', 'name': 'The Next Web'}   

                            author  \
0                    InvestorPlace   
1                Elizabeth Lopatto   
2                             None   
3  prosen@insider.com (Phil Rosen)   
4                       Ivan Mehta   

                                               title  \
0  The Market Has Spoken, and It Says Ethereum Is...   
1   How Ukrainians are fundraising in cryptocurrency   
2  ConsenSys valuation more than doubles to over ...   
3  Bored Ape creator's purchase of CryptoPunks me...   
4  Twitter’s accepting tips in Ethereum — and Dor...   

                                         description  \
0  Technical analysis isn’t a perfect tool, but i...   
1  Millions of dollars of cryptocurrency have flo...   
2  Blockchain technology firm ConsenSys said on T...   
3  Yuga Labs, the creators behind Bored Ape Yacht...   
4  Last night, Twitter announced it’ll allow user...   

                                                 url  \
0  https://investorplace.com/2022/02/the-market-h...   
1  https://www.theverge.com/2022/2/26/22952357/uk...   
2  https://www.reuters.com/technology/consensys-v...   
3  https://markets.businessinsider.com/news/curre...   
4  https://thenextweb.com/news/twitters-tips-eth-...   

                                          urlToImage           publishedAt  \
0  https://images.readwrite.com/wp-content/upload...  2022-02-17T17:10:12Z   
1  https://cdn.vox-cdn.com/thumbor/teEVxppIZ_JTW-...  2022-02-26T20:29:04Z   
2  https://www.reuters.com/pf/resources/images/re...  2022-03-15T15:20:00Z   
3  https://i.insider.com/622f3dd2843ef30018e6ddab...  2022-03-14T13:31:54Z   
4  https://img-cdn.tnwcdn.com/image/hardfork?filt...  2022-02-17T13:43:26Z   

                                             content  
0  Technical analysis isnt a perfect tool, but it...  
1  Illustration by James Bareham / The Verge\r\n\...  
2  March 15 (Reuters) - Blockchain technology fir...  
3  Bored Ape Yacht club creator Yugo Labs has bou...  
4  Last night, Twitter announced itll allow users...

In [8]:
# Pass in the "raw" articles DTO and retrieve a new DataFrame w/ sentiment details
def createDataFrameWithSentiments(articles_dto):
    sentiments = []

    # Iterate through each article and create a new obj that contains the
    # relevant article data (text & date) + sentiment data
    for article in articles_dto["articles"]:
        try:
            text = article["content"]
            date = article["publishedAt"][:10]
            sentiment = analyzer.polarity_scores(text)
            compound = sentiment["compound"]
            pos = sentiment["pos"]
            neu = sentiment["neu"]
            neg = sentiment["neg"]
            
            sentiments.append({
                "text": text,
                "date": date,
                "compound": compound,
                "positive": pos,
                "negative": neg,
                "neutral": neu
                
            })
            
        except AttributeError:
            pass
        
    # Create DataFrame
    df = pd.DataFrame(sentiments)

    # Reorder DataFrame columns
    cols = ["date", "text", "compound", "positive", "negative", "neutral"]
    df = df[cols]

    return df

In [9]:
# Display a few rows of the new data frame
df_btc = createDataFrameWithSentiments(btc_articles_dto)
df_btc.head()

date                                               text  compound  \
0  2022-02-28  LONDON, Feb 28 (Reuters) - Trading volumes bet...   -0.2732   
1  2022-03-04  LONDON, March 4 (Reuters) - Cryptocurrencies h...   -0.1280   
2  2022-02-26  Illustration by James Bareham / The Verge\r\n\...   -0.4588   
3  2022-02-17  Even in cyberspace, the Department of Justice ...    0.7351   
4  2022-03-01  (Kitco News) The escalating Russia-Ukraine cri...   -0.5267   

   positive  negative  neutral  
0     0.000     0.062    0.938  
1     0.000     0.046    0.954  
2     0.000     0.083    0.917  
3     0.147     0.000    0.853  
4     0.047     0.115    0.838

In [10]:
# Describe the Bitcoin Sentiment
df_btc.describe()

compound   positive   negative   neutral
count  20.000000  20.000000  20.000000  20.00000
mean   -0.000170   0.066300   0.055800   0.87795
std     0.492861   0.061578   0.061111   0.06361
min    -0.771300   0.000000   0.000000   0.73900
25%    -0.353350   0.000000   0.000000   0.84475
50%    -0.089800   0.060500   0.054000   0.87050
75%     0.397975   0.092750   0.095750   0.93250
max     0.784500   0.214000   0.187000   1.00000

In [11]:
# Display a few rows of the new data frame
df_eth = createDataFrameWithSentiments(eth_articles_dto)
df_eth.head()

date                                               text  compound  \
0  2022-02-17  Technical analysis isnt a perfect tool, but it...   -0.2498   
1  2022-02-26  Illustration by James Bareham / The Verge\r\n\...   -0.4588   
2  2022-03-15  March 15 (Reuters) - Blockchain technology fir...    0.0000   
3  2022-03-14  Bored Ape Yacht club creator Yugo Labs has bou...    0.5256   
4  2022-02-17  Last night, Twitter announced itll allow users...    0.2263   

   positive  negative  neutral  
0     0.000     0.059    0.941  
1     0.000     0.083    0.917  
2     0.000     0.000    1.000  
3     0.142     0.054    0.803  
4     0.056     0.000    0.944

In [12]:
# Describe the Ethereum Sentiment
df_eth.describe()

compound   positive   negative    neutral
count  20.000000  20.000000  20.000000  20.000000
mean    0.201435   0.083400   0.038050   0.878450
std     0.369257   0.077025   0.035487   0.083807
min    -0.458800   0.000000   0.000000   0.713000
25%    -0.102700   0.000000   0.000000   0.809750
50%     0.214300   0.071000   0.047000   0.896000
75%     0.432825   0.142500   0.059500   0.941750
max     0.834000   0.236000   0.105000   1.000000

### Questions:

Q: Which coin had the highest mean positive score?

A: **Ethereum: 0.083400**

Q: Which coin had the highest compound score?

A: **Ethereum: 0.834000**

Q. Which coin had the highest positive score?

A: **Ethereum: 0.236000**

---

## 2. Natural Language Processing
---
###   Tokenizer

In this section, you will use NLTK and Python to tokenize the text for each coin. Be sure to:
1. Lowercase each word.
2. Remove Punctuation.
3. Remove Stopwords.

In [13]:
from nltk.tokenize import word_tokenize, sent_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer, PorterStemmer
from string import punctuation
import re

In [14]:
# Instantiate the lemmatizer
lemmatizer = WordNetLemmatizer()

# Create a list of stopwords
sw = set(stopwords.words('english'))

# Expand the default stopwords list if necessary
sw = sw.union({'reuters', 'illustration', 'com'})

In [17]:
# Complete the tokenizer function
def tokenizer(text):
    """Tokenizes text."""
    
    # Remove the punctuation from text
    regex = re.compile("[^a-zA-Z ]")
    text = regex.sub('', text)
   
    # Create a tokenized list of the words
    tokens_re = word_tokenize(text)
    
    # Lemmatize words into root words
    tokens_lemmatized = [lemmatizer.lemmatize(token) for token in tokens_re] 
   
    # Convert the words to lowercase & remove the stop words
    tokens = [token.lower() for token in tokens_lemmatized if token.lower() not in sw]
    
    return tokens

In [19]:
# Create a new tokens column for Bitcoin
df_btc['btc_tokens'] = df_btc.apply(lambda row: (tokenizer(row.text)),axis=1)
df_btc.head()

date                                               text  compound  \
0  2022-02-28  LONDON, Feb 28 (Reuters) - Trading volumes bet...   -0.2732   
1  2022-03-04  LONDON, March 4 (Reuters) - Cryptocurrencies h...   -0.1280   
2  2022-02-26  Illustration by James Bareham / The Verge\r\n\...   -0.4588   
3  2022-02-17  Even in cyberspace, the Department of Justice ...    0.7351   
4  2022-03-01  (Kitco News) The escalating Russia-Ukraine cri...   -0.5267   

   positive  negative  neutral  \
0     0.000     0.062    0.938   
1     0.000     0.046    0.954   
2     0.000     0.083    0.917   
3     0.147     0.000    0.853   
4     0.047     0.115    0.838   

                                          btc_tokens  
0  [london, feb, trading, volume, russian, rouble...  
1  [london, march, cryptocurrencies, close, headl...  
2  [james, bareham, verge, million, ha, donated, ...  
3  [even, cyberspace, department, justice, able, ...  
4  [kitco, news, escalating, russiaukraine, crisi...

In [20]:
# Create a new tokens column for Ethereum
df_eth['eth_tokens'] = df_eth.apply(lambda row: (tokenizer(row.text)),axis=1)
df_eth.head()

date                                               text  compound  \
0  2022-02-17  Technical analysis isnt a perfect tool, but it...   -0.2498   
1  2022-02-26  Illustration by James Bareham / The Verge\r\n\...   -0.4588   
2  2022-03-15  March 15 (Reuters) - Blockchain technology fir...    0.0000   
3  2022-03-14  Bored Ape Yacht club creator Yugo Labs has bou...    0.5256   
4  2022-02-17  Last night, Twitter announced itll allow users...    0.2263   

   positive  negative  neutral  \
0     0.000     0.059    0.941   
1     0.000     0.083    0.917   
2     0.000     0.000    1.000   
3     0.142     0.054    0.803   
4     0.056     0.000    0.944   

                                          eth_tokens  
0  [technical, analysis, isnt, perfect, tool, may...  
1  [james, bareham, verge, million, ha, donated, ...  
2  [march, blockchain, technology, firm, consensy...  
3  [bored, ape, yacht, club, creator, yugo, labs,...  
4  [last, night, twitter, announced, itll, allow,...

---

### NGrams and Frequency Analysis

In this section you will look at the ngrams and word frequency for each coin. 

1. Use NLTK to produce the n-grams for N = 2. 
2. List the top 10 words for each coin. 

In [15]:
from collections import Counter
from nltk import ngrams

In [16]:
# Generate the Bitcoin N-grams where N=2
# YOUR CODE HERE!

In [17]:
# Generate the Ethereum N-grams where N=2
# YOUR CODE HERE!

In [18]:
# Function token_count generates the top 10 words for a given coin
def token_count(tokens, N=3):
    """Returns the top N tokens from the frequency count"""
    return Counter(tokens).most_common(N)

In [19]:
# Use token_count to get the top 10 words for Bitcoin
# YOUR CODE HERE!

In [20]:
# Use token_count to get the top 10 words for Ethereum
# YOUR CODE HERE!

---

### Word Clouds

In this section, you will generate word clouds for each coin to summarize the news for each coin

In [21]:
from wordcloud import WordCloud
import matplotlib.pyplot as plt
plt.style.use('seaborn-whitegrid')
import matplotlib as mpl
mpl.rcParams['figure.figsize'] = [20.0, 10.0]

In [22]:
# Generate the Bitcoin word cloud
# YOUR CODE HERE!

In [23]:
# Generate the Ethereum word cloud
# YOUR CODE HERE!

---
## 3. Named Entity Recognition

In this section, you will build a named entity recognition model for both Bitcoin and Ethereum, then visualize the tags using SpaCy.

In [24]:
import spacy
from spacy import displacy

In [25]:
# Download the language model for SpaCy
# !python -m spacy download en_core_web_sm

In [26]:
# Load the spaCy model
nlp = spacy.load('en_core_web_sm')

---
### Bitcoin NER

In [27]:
# Concatenate all of the Bitcoin text together
# YOUR CODE HERE!

In [28]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [29]:
# Render the visualization
# YOUR CODE HERE!

In [30]:
# List all Entities
# YOUR CODE HERE!

---

### Ethereum NER

In [31]:
# Concatenate all of the Ethereum text together
# YOUR CODE HERE!

In [32]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [33]:
# Render the visualization
# YOUR CODE HERE!

In [34]:
# List all Entities
# YOUR CODE HERE!

---